feat(component): add Worksheet component

jorgemoya opened the pull request:
## Beautiful Preview
Here it is people (or at least the foundation).

Be amazed by the beauty:

![Kapture 2021-04-08 at 11 26 57](https://user-images.githubusercontent.com/196129/114062751-7b1fe380-985d-11eb-9403-b68208770084.gif)

Wow very nice.

## What

The idea of this PR is to review the foundation for the Worksheet component but without exposing/releasing it to the public. This is due to the complexity of this component. There are still a lot of features we need to implement before public release.

Proposed API:
```ts
const columns: WorksheetColumn[] = [
  { hash: 'productName', header: 'Product name' },
  { hash: 'otherField', header: 'Other field', type: 'number', validation: (value) => value !== 2 },
];

const items = [
  {
    productName: 'Epic shoes',
    otherField: 1,
  },

  {
    productName: 'Not so nice shoes',
    otherField: 2,
  },
]

<Worksheet columns={columns} items={items} onChange={handleChange} onErrors={handleErrors} />;
```

## How
I'm using the [zustand](https://github.com/pmndrs/zustand) library to handle state to prevent unnecessary rerendering.

I'm trying to think ahead and many of the functions I build support multiple cell parameters even thought we're currently constrained to a single cell selection.

## Features

- Only Text/Number editor for now
- Cell visual selection
- Validation
- Returns edited cells on change
- Returns invalid cells on first pass and after a change

## Notes

- As you can see, there are a lot of missing features expected from a Worksheet component (click-n-drag cells, select all, undo/redo, row/column select, adjustable column width, etc...), but it is WIP in progress.
- No keyboard listeners other than enter and escape when editing a cell.
- No mouse multi selection.
- I haven't thought about accessibility.
- For launch we definitely need at least two other field types `select` & `checkbox`.
- Virtualization is planned.
- No docs.

jorgemoya commented on packages/big-design/src/components/Worksheet/hooks/spec.tsx:
```diff
@@ -0,0 +1,109 @@
+  });
+});
+
+// jest.mock('./useUpdateItems', () => jest.requireActual('./useUpdateItems'));
+
+//TODO: fix this test
+// describe('useUpdateItems', () => {
+//   // beforeEach(() => {
+//   //   jest.unmock('./useUpdateItems');
+//   // });
+
+//   test('throws if is not wrapped in UpdateItemsProvider', () => {
+//     const { result } = renderHook(() => useUpdateItems());
+
+//     expect(result.error.message).toContain('UpdateItemsProvider');
+//   });
+// });
```
I couldn't figure out how to add this test since I mock `useUpdateItems` for the other component above.

jorgemoya replied:
I feel like this test isn't really necessary. Do you feel strongly we need it?

chanceaclark replied:
Eh, get rid of it then.

chanceaclark commented on packages/big-design/src/components/Worksheet/types.ts:
```diff
@@ -0,0 +1,21 @@
+  columnIndex: number;
+  hash: string;
+  rowIndex: number;
+  value: string | number;
```
Let's see if we can infer this type through a generic. I took a quick look, and you might explore this:
```ts
interface Product {
    productName: string;
    categories: string;
    otherField: string;
    otherField2: string;
    otherField3: number;
}

interface Worksheet<Item> {
    items: Item[];
    onChange(items: Array<Cell<Item[keyof Item]>>): void;
}

interface Cell<Value> {
    value: Value;
}

const ws: Worksheet<Product> = {
    items: [],
    onChange: items => {
        console.log(items);
    },
};
```
![Screen Shot 2021-04-05 at 16 09 41](https://user-images.githubusercontent.com/10539418/113627599-5fcc9280-9629-11eb-893f-b4663308c638.png)

jorgemoya replied:
This is super gucci but the problem I see is that I use the `Cell` type in a lot of places. What would be a good solution for that?

chanceaclark replied:
Yeah, `unknown` would be the best bet. Is there a problem with that approach?

chanceaclark commented on packages/big-design/src/components/Worksheet/utils.ts:
```diff
@@ -0,0 +1,20 @@
+
+export const mergeCells = (oldCells: Cell[], newCells: Cell[]) =>
+  newCells.reduce(
+    (accum, cell) => {
+      const index = oldCells.findIndex((c) => c.rowIndex === cell.rowIndex && c.columnIndex === cell.columnIndex);
```
Same with above, could we make `c` named more developer friendly?

chanceaclark commented on packages/big-design/src/components/Worksheet/styled.tsx:
```diff
@@ -0,0 +1,22 @@
+  border: ${({ theme }) => `${remCalc(0.5)} solid ${theme.colors.secondary30}`};
+  box-sizing: border-box;
+  color: ${({ theme }) => theme.colors.secondary60};
+  font-weight: 600; //TODO: fix?
```
`semiBold`?

chanceaclark commented on packages/big-design/src/components/Worksheet/styled.tsx:
```diff
@@ -0,0 +1,22 @@
+import { theme as defaultTheme, remCalc } from '@bigcommerce/big-design-theme';
```
I think we should be trying to use `remCalc` from the `theme` object instead of importing it 🤔  Did you see any other usage like this?

jorgemoya replied:
Honestly I don't remember using it from the `theme` object in BigDesign before, I might need to take a look at other components at a later time.

chanceaclark reviewed (commented):
Initial review, once we get it in a gucci state, I'll ♻️  again.

jorgemoya commented on packages/big-design/src/components/Worksheet/RowStatus/RowStatus.tsx:
```diff
@@ -0,0 +1,19 @@
+
+import { Status } from './styled';
+
+export interface RowStatusProps {
+  rowIndex: number;
+}
```
In this example I dont need to `export` it anymore (fixed), but do you mean move all types to the parent `./types` file?

My idea was to inline `ComponentProps` or local types and add only "global" type to `/Worksheet/types.ts` like `Cell`.

chanceaclark replied:
>do you mean move all types to the parent ./types file?

Sort of yeah. With the Tree component types, I sort of mix-n-matched where I threw them. I don't know what the best approach is, but was wondering your thoughts on the matter?

jorgemoya replied:
I like the idea of keeping `ComponentProps` types inline and the rest in `./types.ts`

jorgemoya commented on packages/big-design/src/components/Worksheet/editors/TextEditor/TextEditor.tsx:
```diff
@@ -0,0 +1,39 @@
+
+export interface TextEditorProps {
+  cell: Cell;
+  handleBlur(): void;
+  handleKeyDown: EditableCellKeyDown;
```
You mean change them to `onBlur` and `onKeyDown`?

chanceaclark replied:
Yep!

jorgemoya commented on packages/big-design/src/components/Worksheet/editors/TextEditor/styled.ts:
```diff
@@ -0,0 +1,27 @@
+  ${({ isEdited }) =>
+    isEdited &&
+    css`
+      background-color: ${({ theme }) => theme.colors.warning20};
```
What designs you looking at? The ones I see at this [figma](https://www.figma.com/file/zcJRWjXwgLGRcpxKrdFkwZ/Starter-App-Outline?node-id=1994%3A11015) file has them as warning20.

jordan-massingill commented on packages/big-design/src/components/Worksheet/Cell/Cell.tsx:
```diff
@@ -0,0 +1,103 @@
+  const isSelected = useStore(
+    useMemo(
+      () => (state) =>
+        state.selectedCells.reduce(
```
I like this use of reduce, but jw is there a reason not to use `Array.some()` here instead?

```suggestion
        state.selectedCells.some(selectedCell => selectedCell.columnIndex === cell.columnIndex && selectedCell.rowIndex === cell.rowIndex)
```

jorgemoya replied:
Good call

jordan-massingill commented on packages/big-design/src/components/Worksheet/RowStatus/RowStatus.tsx:
```diff
@@ -0,0 +1,19 @@
+  const isSelected = useStore(useMemo(() => (state) => state.selectedRows.includes(rowIndex), [rowIndex]));
+
+  const isInvalid = useStore(
+    useMemo(() => (state) => Boolean(state.invalidCells.find((cell) => cell.rowIndex === rowIndex)), [rowIndex]),
```
🍹  This line may benefit from a `.some()` as well since it already returns a boolean value.

rtalvarez commented on packages/big-design/src/components/Worksheet/Worksheet.tsx:
```diff
@@ -0,0 +1,74 @@
+    }
+  }, [invalidCells, onErrors]);
+
+  const renderHeaders = useMemo(
```
🍹 nitpicky comment, since this const holds an actual value and not a callback, i'd suggest renaming this to `renderedHeaders` 

(same for below)

rtalvarez commented on packages/big-design/src/components/Worksheet/hooks/spec.tsx:
```diff
@@ -0,0 +1,91 @@
+      result.current.handleDoubleClick();
+      result.current.handleKeyDown(
+        ({
+          currentTarget: 1,
```
curious, what are these `currentTarget` and `target` fields for in a React.KeyboardEvent? Can we just not supply them?

Also, can we not construct this object using the native API `new KeyboardEvent('Foo', { key: 'Enter' })` ?

jorgemoya replied:
To be honest it's a copy pasta from another test. I can simply remove them since they are not relevant to current test, good catch. I couldn't implement your solution, can you expand on it?

rtalvarez commented on packages/big-design/src/components/Worksheet/styled.tsx:
```diff
@@ -0,0 +1,22 @@
+
+export const Header = styled.th`
+  background-color: ${({ theme }) => theme.colors.secondary10};
+  border: ${({ theme }) => `${theme.helpers.remCalc(0.5)} solid ${theme.colors.secondary30}`};
```
🍹  I wonder if we need a "tableBorder" internal theme prop or something to provide this styling, it'll be easier to enforce, maintain and use

rtalvarez reviewed (commented):
looking good

chanceaclark commented on packages/big-design/src/components/Worksheet/Cell/Cell.tsx:
```diff
@@ -0,0 +1,101 @@
+  validation: WorksheetColumn<Value>['validation'];
+}
+
+// We can't know what value type a cell has
+export const Cell: React.FC<CellProps<any>> = memo(({ columnIndex, hash, rowIndex, type, value, validation }) => {
```
You can still use generics here:
```suggestion
const InternalCell = <T extends unknown>({
  columnIndex,
  hash,
  rowIndex,
  type,
  value,
  validation,
}: CellProps<T>): React.ReactElement<CellProps<T>> => {
```

And the export will be:
```tsx
export const Cell = typedMemo(InternalCell);
```

I would still try and type these with generics if you can.

chanceaclark commented on packages/big-design/src/components/Worksheet/Row/Row.tsx:
```diff
@@ -0,0 +1,32 @@
+import { RowStatus } from '../RowStatus';
+import { WorksheetColumn } from '../types';
+
+interface RowProps {
+  columns: WorksheetColumn<unknown>[];
+  rowIndex: number;
+}
+
+export const Row: React.FC<RowProps> = memo(({ columns, rowIndex }) => {
```
Still think you can use generics here too.

chanceaclark commented on packages/big-design/src/components/Worksheet/Row/Row.tsx:
```diff
@@ -0,0 +1,32 @@
+          hash={column.hash}
+          key={`${rowIndex}-${columnIndex}`}
+          rowIndex={rowIndex}
+          type={column.type || 'text'}
```
```suggestion
          type={column.type ?? 'text'}
```

chanceaclark commented on packages/big-design/src/components/Worksheet/hooks/spec.tsx:
```diff
@@ -0,0 +1,87 @@
+}));
+
+describe('useEditableCells', () => {
+  const cell: Cell<unknown> = { rowIndex: 0, columnIndex: 0, type: 'text', hash: 'category', value: 'Text' };
```
Shouldn't this be `string`?
```suggestion
  const cell: Cell<string> = { rowIndex: 0, columnIndex: 0, type: 'text', hash: 'category', value: 'Text' };
```

chanceaclark reviewed (commented):
Few more comments but looks really good man

jorgemoya commented:
Per offline feedback, I have now added a check to only pass value of type `string | number` to `Cells`. So I updated the types to reflect this change.

Most noticeably, `validation` now should always return a `string | number` as the type of the value prop.

Which now makes me think if instead of adding `generics`, should we just expect the value to be `string | number` since we enforce it now anyway?

chanceaclark reviewed (approved):
Looks good to me! Great job on this!
